### PR TITLE
add avrowriter

### PR DESCRIPTION
**Pull request by alice2008**
Add avrowriter servers as the counter part of `AvroDataReader`. @joshvfleming @ashelkovnykov

### CODE REVIEW

File: photon-avro-schemas/src/main/avro/ResponsePredictionAvro.avsc [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  }, {
+    "name" : "weight",
+    "type" : "double",
+    "default": 1.0
+  }, {
+    "name" : "offset",
+    "type" : "double",
+    "default": 0.0
+  }, {
+    "name" : "randomTestField",
```
Comment (joshvfleming) [inline photon-avro-schemas/src/main/avro/ResponsePredictionAvro.avsc]: We shouldn't need this field.

File: photon-client/src/integTest/scala/com/linkedin/photon/ml/data/avro/AvroDataWriterIntegTest.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    val fs = FileSystem.get(sc.hadoopConfiguration)
+    val files = fs.listStatus(new Path(outputPath)).filter(_.getPath.getName.startsWith("part"))
+    assertEquals(files.length, numPartitions)
+  }
+}
+
+object AvroDataWriterIntegTest {
+  private val inputDir = getClass.getClassLoader.getResource("GameIntegTest/input").getPath
+  private val inputPath = new Path(inputDir, "train")
+  private val outputPath = "/tmp/AvroDataWriter/"
```
Comment (joshvfleming) [inline photon-client/src/integTest/scala/com/linkedin/photon/ml/data/avro/AvroDataWriterIntegTest.scala]: Use the `TestTemplateWithTmpDir` mixin to generate a temporary folder instead. It generates the folder in a more OS-portable way, guarantees that access to the folder is thread-safe, and also cleans up the temporary folder after the tests finish.

File: photon-client/src/integTest/scala/com/linkedin/photon/ml/data/avro/AvroDataWriterIntegTest.scala [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    assertTrue(df.columns.contains(responseColumn))
+    assertEquals(df.count, 34810)
+    assertTrue(indexMapLoadersMap.contains(featureColumn))
+
+    val indexMapLoader = indexMapLoadersMap(featureColumn)
+    AvroDataWriter.saveAvro(sc, df, outputPath, indexMapLoader, responseColumn, featureColumn)
+
+    val fs = FileSystem.get(sc.hadoopConfiguration)
+    val files = fs.listStatus(new Path(outputPath)).filter(_.getPath.getName.startsWith("part"))
+    assertEquals(files.length, numPartitions)
```
Comment (joshvfleming) [inline photon-client/src/integTest/scala/com/linkedin/photon/ml/data/avro/AvroDataWriterIntegTest.scala]: Let's also read the DataFrame back after saving it and do some basic sanity checks, to make sure the full circle works as expected.

File: photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  }
+
+  /**
+   * Build a list of Avro Feature instances for the given list [[Vector]] and [[IndexMap]]
+   *
+   * @param vector The extracted feature in [[Vector]] for a particular training instance
+   * @param indexMap The reverse index map from feature to index
+   * @return A list of Avro Feature instances built from the vector
+   */
+  protected[data] def buildAvroFeatures(vector: Vector, indexMap: Map[Int, String]): java.util.List[FeatureAvro] = {
```
Comment (joshvfleming) [inline photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala]: You can use Scala data strucures instead here, and then convert to Java at the end when needed, e.g.:

```scala
import scala.collection.JavaConverters._
List(1,2,3).asJava
```

File: photon-client/src/integTest/scala/com/linkedin/photon/ml/data/avro/AvroDataWriterIntegTest.scala (hunk 3 above)
Comment (alice2008) [reply photon-client/src/integTest/scala/com/linkedin/photon/ml/data/avro/AvroDataWriterIntegTest.scala]: done

File: photon-client/src/integTest/scala/com/linkedin/photon/ml/data/avro/AvroDataWriterIntegTest.scala (hunk 2 above)
Comment (alice2008) [reply photon-client/src/integTest/scala/com/linkedin/photon/ml/data/avro/AvroDataWriterIntegTest.scala]: done

File: photon-avro-schemas/src/main/avro/ResponsePredictionAvro.avsc (hunk 1 above)
Comment (alice2008) [reply photon-avro-schemas/src/main/avro/ResponsePredictionAvro.avsc]: done

File: photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+   * @param responseColumn The response column name in df
+   * @param featureColumn The feature column name in df
+   */
+  def saveAvro(
+      sc: SparkContext,
+      df: DataFrame,
+      outputPath: String,
+      indexMapLoader: IndexMapLoader,
+      responseColumn: String,
+      featureColumn: String): Unit = {
```
[the remark this answers is not in the document]
Comment (alice2008) [reply photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala]: add a TODO

File: photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+   *
+   * @param sc The Spark Context
+   * @param df The DataFrame
+   * @param outputPath The output path to store the avro files
+   * @param indexMapLoader The IndexMapLoader store feature to index information
+   * @param responseColumn The response column name in df
+   * @param featureColumn The feature column name in df
+   */
+  def saveAvro(
+      sc: SparkContext,
```
[the remark this answers is not in the document]
Comment (alice2008) [reply photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala]: done

File: photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+          .build()
+      }
+    }
+
+    // Write the converted dataset back to HDFS
+    // If the output paths already exist, remove them
+    val fs = FileSystem.get(sc.hadoopConfiguration)
+    val output = new Path(outputPath)
+    if (fs.exists(output)) {
+      fs.delete(output, true)
```
[the remark this answers is not in the document]
Comment (alice2008) [reply photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala]: done. add `overwrite` field

File: photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+          case None =>
+            responseType = Some(r.schema(responseColumn).dataType)
+            if (hasOffset) {
+              offsetType = Some(r.schema("offset").dataType)
+            }
+            if (hasWeight) {
+              weightType = Some(r.schema("weight").dataType)
+            }
+          case Some(_) =>
+        }
```
[the remark this answers is not in the document]
Comment (alice2008) [reply photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala]: done

File: photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+   * @return A double in this field
+   */
+  protected[data] def getValueAsDouble(row: Row, fieldName: String, dataType: DataType): Double = {
+
+    dataType match {
+      case IntegerType => row.getAs[Int](fieldName).toDouble
+      case LongType => row.getAs[Long](fieldName).toDouble
+      case FloatType => row.getAs[Float](fieldName).toDouble
+      case BooleanType => if (row.getAs[Boolean](fieldName)) 1.0D else 0.0D
+      case DoubleType => row.getAs[Double](fieldName)
```
[the remark this answers is not in the document]
Comment (alice2008) [reply photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala]: done. This is much neat! thanks.

File: photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+    val hasOffset = columns.contains("offset")
+    val hasWeight = columns.contains("weight")
+
+    // Build reverse index map from feature to index and broadcast
+    // When feature space is large, it is valid to iterate sparse vector space
+    // instead of feature space
+    val indexMap = indexMapLoader.indexMapForDriver()
+    val reverseIndexMap = indexMap.map(_.swap)
+    val broadcastReverseMap = sc.broadcast(reverseIndexMap)
```
[the remark this answers is not in the document]
Comment (alice2008) [reply photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala]: thanks for the suggestions! done.

File: photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    require(columns.contains(featureColumn), s"There must be a $featureColumn column present in dataframe")
+
+    val hasOffset = columns.contains("offset")
+    val hasWeight = columns.contains("weight")
+
+    // Build reverse index map from feature to index and broadcast
+    // When feature space is large, it is valid to iterate sparse vector space
+    // instead of feature space
+    val indexMap = indexMapLoader.indexMapForDriver()
+    val reverseIndexMap = indexMap.map(_.swap)
```
[the remark this answers is not in the document]
Comment (alice2008) [reply photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala]: done.

File: photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+   * Write the DataFrame into avro records using the given indexMapLoader
+   *
+   * @param sc The Spark Context
+   * @param df The DataFrame
+   * @param outputPath The output path to store the avro files
+   * @param indexMapLoader The IndexMapLoader store feature to index information
+   * @param responseColumn The response column name in df
+   * @param featureColumn The feature column name in df
+   */
+  def saveAvro(
```
[the remark this answers is not in the document]
Comment (alice2008) [reply photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala]: done.

File: photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala (hunk 4 above)
Comment (alice2008) [reply photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala]: done

**Pull request merged**